sycl/include/sycl/accessor.hpp

```diff
@@ -2215,8 +2219,78 @@ class __SYCL_SPECIAL_CLASS accessor<DataT, Dimensions, AccessMode,
     return local_ptr<DataT>(getQualifiedPtr());
   }
 
-  bool operator==(const accessor &Rhs) const { return impl == Rhs.impl; }
-  bool operator!=(const accessor &Rhs) const { return !(*this == Rhs); }
+  bool operator==(const local_accessor_base &Rhs) const {
+    return impl == Rhs.impl;
+  }
+  bool operator!=(const local_accessor_base &Rhs) const {
+    return !(*this == Rhs);
+  }
+};
+
+template <typename DataT, int Dimensions, access::mode AccessMode,
+          access::placeholder IsPlaceholder>
+class __SYCL_SPECIAL_CLASS accessor<DataT, Dimensions, AccessMode,
+                                    access::target::local, IsPlaceholder>
+    : public local_accessor_base<DataT, Dimensions, AccessMode, IsPlaceholder> {
+
+  using local_acc =
+      local_accessor_base<DataT, Dimensions, AccessMode, IsPlaceholder>;
+
+  // Use base classes constructors
+  using local_acc::local_acc;
+
+#ifdef __SYCL_DEVICE_ONLY__
+
+  // __init needs to be defined within the class not through inheritance.
+  // Map this function to inherited func.
+  void __init(typename local_acc::ConcreteASPtrType Ptr,
+              range<local_acc::AdjustedDim> AccessRange,
+              range<local_acc::AdjustedDim> range,
+              id<local_acc::AdjustedDim> id) {
+    local_acc::__init(Ptr, AccessRange, range, id);
+  }
+
+public:
+  // Default constructor for objects later initialized with __init member.
+  accessor() {
+    local_acc::impl = detail::InitializedVal<local_acc::AdjustedDim,
+                                             range>::template get<0>();
+  }
+
+#endif
+};
+
+template <typename DataT, int Dimensions = 1>
+class __SYCL_SPECIAL_CLASS local_accessor
+    : public local_accessor_base<DataT, Dimensions, access::mode::read_write,
+                                 access::placeholder::false_t> {
```

Comment:
Would it be possible to rename `local_accessor_base` to `local_accessor` and just have the specialization above inherit from that? Means that we won't have to flatten it when we remove the above specialization.

Reply:
So I am not sure that `local_accessor_base` can be renamed `local_accessor` for a few reasons, one of which is you cannot use default template arguments when partially specializing. In this case, the dimension has a default argument. There may be other issues around declarations for `multi_ptr` and having a specialization inherit from itself. Perhaps I am misunderstanding the approach you are recommending. Please clarify if I have. Additionally, I don't believe the name change from `local_accessor_base` to `local_accessor` will cause too much hassle when flattening.

Reply:
Fair. I am okay with it as-is.